### DISCUSSION TITLE
Workspace Editor

### DISCUSSION THREAD

**Discussion by N0cynym** [jglim/CaesarSuite#21] [Ideas]
Now Diogenes is a very powerful software with some unique tools.
As [Sergey](https://github.com/jglim/CaesarSuite/discussions/1#discussioncomment-240414) was asking for a frame to work with fault codes I was thinking about the capability how a user can create his own customised work bench.

In **Vediamo** the frame is not really dynamically. If you'e a professionel you can build you're bench config file `.vsb` and be able to create measurement files `.mwg` or service groups `.vsg` and you can also create a quick test or xml routine.

**DTS** offers a more dynamic and standardized option to create a work bench.

_I'm not really sure if my thoughts are going to much away from the main idea of CaesarSuite. So please excuse me._

**Well here is my idea:**

- A new window to build a xml based work bench file (like DTS hmi modules to create and load a work bench).
- Let the user drag and drop services from main Diogenes to the work bench (and also edit them).
- Offering VC as a module like in DTS and build from that point all new features as modules

Now at this point I will stop.
Please let me know if this idea is interesting for you.
If not fell free to delete this discussion.
Else I will explain it more detailed and bring into our wiki all the knowledge to the standard section.

**Comment by jglim** [comment]
I think it is definitely interesting. The key concern is the additional effort required to modify every extra feature whenever API changes are required, since there's a significant amount of content that hasn't been fixed up yet. Take DTC for example: it resides in ECU.cs, and I have not started work on it. It also has additional references to the unimplemented Environmental data.

My current plan is to stabilize Caesar/Diogenes enough, and when internal functions do not change as often, it will be a better time to plan these additional features. This discussion is good and should remain, so that at a later time it can be reviewed and maybe worked on.

**Comment by andcarand** [comment]
Hello everyone! thanks for good job!
I am getting an error while trying to read errors. Is it possible to fix it somehow?
![erroe dtc](https://user-images.githubusercontent.com/82184815/114104786-37b48c00-98d4-11eb-9a6c-d44b5ff5ce66.jpeg)

**Reply by jglim** [reply]
Hello, thanks for raising this issue. 
I have applied a bugfix, could you try this build and see if it fixes the issue? 

[CaesarSuite_dbg_2021-04-09.zip](https://github.com/jglim/CaesarSuite/files/6282796/CaesarSuite_dbg_2021-04-09.zip)

**Reply by andcarand** [reply]
Good day. Thank you!
This fixed the application bug.
But now there are no errors in any blocks. 
If you connect Vediamo, then there are errors. 
checked ezs166, ezs212, zgw211, zgw164.

**Reply by jglim** [reply]
That's unusual, could you assist me in this:

- Connect, open DTCs the same way as you have done earlier
- Open the Trace window (`Tools → Show Trace` or `Alt+Q`)
- Copy out the entire log, or save to file
- Paste or upload the log here

Just EIS166 should be enough, I will investigate from there.

From your earlier screenshot, I was expecting to see these DTCS for EIS166:

```
D03D2A U103D2A : Die Fahrwerk-CAN-2-Kommunikation hat Funktionsstörung. Keine Signaländerung
D0322A U10322A : Die Fahrwerk-CAN-Kommunikation hat Funktionsstörung. Keine Signaländerung
C0192A U00192A : Die Innenraum-CAN-Kommunikation hat Funktionsstörung. Keine Signaländerung
D18200 U118200 : Das Fahrwerk-CAN-BUS-Netzwerkmanagement ist instabil.
D18000 U118000 : Das Innenraum-CAN-BUS-Netzwerkmanagement ist instabil.
D18300 U118300 : Das Diagnose-CAN-BUS-Netzwerkmanagement ist instabil.
C12287 U012287 : Die Kommunikation mit dem ESP hat eine Funktionsstörung (Timeout).
D18100 U118100 : Das Fahrwerk-CAN-2-BUS-Netzwerkmanagement ist instabil.
```

This is definitely a bug in the Diogenes DTC parser; thank you for your patience.

**Reply by andcarand** [reply]
I'll be glad to help! any tests, logs, checks.
this is how it looks on the screen.
in the text file logs.

![Screenshot (20210409-171106)](https://user-images.githubusercontent.com/82184815/114195755-6c1e5b80-9959-11eb-9336-59c670331d14.jpg)

[Trace_20210409_1724.txt](https://github.com/jglim/CaesarSuite/files/6286446/Trace_20210409_1724.txt)

**Reply by jglim** [reply]
Thanks for the high-quality logs and screenshot. I've applied another round of fixes, and with your logs, hopefully you should see something like this now:
![image](https://user-images.githubusercontent.com/1116555/114203845-1cf11e80-998b-11eb-8bd2-eb9cfa9ca074.png)

[CaesarSuite_dbg_2021-04-09-B.zip](https://github.com/jglim/CaesarSuite/files/6286747/CaesarSuite_dbg_2021-04-09-B.zip)

**Reply by andcarand** [reply]
very fast! thanks for this! right now I will test and write.

**Comment by andcarand** [comment]
Excellent! It works! Thank you!!!
I will test further. If you have any questions, I will write.
Is error reading available only for UDS? read eeprom the same way?
Thank you!!!

![2021-04-09-18-39-41](https://user-images.githubusercontent.com/82184815/114205907-76ddee00-9963-11eb-95c7-a3b9d66242db.png)

**Reply by jglim** [reply]
Glad to hear that it works! DTC is currently for UDS only, since the command and data layout is most likely different for pre-UDS (e.g. KW2C3PE). I don't have any of those devices around, nor have I read the specifications. I might create an issue for this at a later date to collect traces and work on it.

As for EEPROM (referring to the UDS hex editor?), it depends on the UDS 0x23/0x3D _Read/Write Memory By Address_ and I do not know if there is an equivalent for pre-UDS devices.

**Reply by andcarand** [reply]
OK. i will test.
compared the commands for reading KI 211 and IC204 after initializing the protocol and holding an active session, the commands do not differ. But I'll check again what protocol sd connect is connected to.

**Comment by potap1966** [comment]
Good day.

I tried to read the flash memory of the VGS4NAG2 (UDS) block with the UDS 0x23 commands.
The block was previously unlocked for programming (seed 05).
As a result, I get zeros (the answer is always 7F 23 11).
This block cannot be read with UDS 0x23?

**Reply by Feezex** [reply]
Different ecu's have its own OTP areas, thus why not all memory may be accessible even with Security services. You must knew exact offset that can be read out.
